# Exploratory Data Analysis

The goal of this project is to evaluate the impact of different first-order discount levels on new user purchase behavior. We will simulate a controlled A/B test with four discount groups and analyze first-order GMV (Gross Merchandise Value). This notebook focuses on exploratory data analysis (EDA), cleaning, and preparing the dataset for A/B test analysis.

In [3]:
# Core libraries
import pandas as pd
import numpy as np

# Visualization
import matplotlib.pyplot as plt

# Display settings
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)

## Data Loading & Inspection

In [6]:
# Load raw order history
df = pd.read_csv('/Users/Cathaml/Desktop/order_history_kaggle_data.csv', header=0)

# Preview
display(df.head())

# Data summary
df.info()
df.describe()

Restaurant ID Restaurant name   Subzone       City    Order ID  \
0       20320607           Swaad  Sector 4  Delhi NCR  6168884918   
1       20320607           Swaad  Sector 4  Delhi NCR  6170707559   
2       20320607           Swaad  Sector 4  Delhi NCR  6169375019   
3       20320607           Swaad  Sector 4  Delhi NCR  6151677434   
4       20320607           Swaad  Sector 4  Delhi NCR  6167540897   

               Order Placed At Order Status         Delivery Distance  \
0  11:38 PM, September 10 2024    Delivered  Zomato Delivery      3km   
1  11:34 PM, September 10 2024    Delivered  Zomato Delivery      2km   
2  03:52 PM, September 10 2024    Delivered  Zomato Delivery     <1km   
3  03:45 PM, September 10 2024    Delivered  Zomato Delivery      2km   
4  03:04 PM, September 10 2024    Delivered  Zomato Delivery      2km   

                                      Items in order Instructions  \
0  1 x Grilled Chicken Jamaican Tender, 1 x Grill...          NaN   
1  1 x Peri Peri Fries, 1 x Fried Chicken Angara ...          NaN   
2              1 x Bone in Peri Peri Grilled Chicken          NaN   
3  1 x Fried Chicken Ghostbuster Tender, 1 x Anga...          NaN   
4  1 x Peri Peri Krispers, 1 x Fried Chicken Anga...          NaN   

   Discount construct  Bill subtotal  Packaging charges  \
0  40% off upto Rs.80          715.0              31.75   
1     Flat Rs.175 off         1179.0              50.20   
2  40% off upto Rs.80          310.0              11.50   
3  40% off upto Rs.80          620.0              27.00   
4  40% off upto Rs.80          584.0              25.20   

   Restaurant discount (Promo)  \
0                         80.0   
1                        175.0   
2                         80.0   
3                         80.0   
4                         80.0   

   Restaurant discount (Flat offs, Freebies & others)  Gold discount  \
0                                                0.0             0.0   
1                                                0.0             0.0   
2                                                0.0             0.0   
3                                                0.0             0.0   
4                                                0.0             0.0   

   Brand pack discount    Total  Rating Review  \
0                  0.0   666.75     NaN    NaN   
1                  0.0  1054.20     NaN    NaN   
2                  0.0   241.50     NaN    NaN   
3                  0.0   567.00     4.0    NaN   
4                  0.0   529.20     NaN    NaN   

  Cancellation / Rejection reason  Restaurant compensation (Cancellation)  \
0                             NaN                                     NaN   
1                             NaN                                     NaN   
2                             NaN                                     NaN   
3                             NaN                                     NaN   
4                             NaN                                     NaN   

   Restaurant penalty (Rejection)  KPT duration (minutes)  \
0                             NaN                   18.35   
1                             NaN                   16.95   
2                             NaN                   14.05   
3                             NaN                   19.00   
4                             NaN                   15.97   

   Rider wait time (minutes) Order Ready Marked Customer complaint tag  \
0                       11.6          Correctly                    NaN   
1                        3.6          Correctly                    NaN   
2                       12.2          Correctly                    NaN   
3                        3.3          Correctly                    NaN   
4                        1.0          Correctly                    NaN   

                                         Customer ID  
0  5d6c2b96db963098bc69768bea504c8bf46106a8a5178e...  
1  0781815deb4a10a574e9fee4fa0b86b074d4a0b36175d5...  


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21321 entries, 0 to 21320
Data columns (total 29 columns):
 #   Column                                              Non-Null Count  Dtype  
---  ------                                              --------------  -----  
 0   Restaurant ID                                       21321 non-null  int64  
 1   Restaurant name                                     21321 non-null  object 
 2   Subzone                                             21321 non-null  object 
 3   City                                                21321 non-null  object 
 4   Order ID                                            21321 non-null  int64  
 5   Order Placed At                                     21321 non-null  object 
 6   Order Status                                        21321 non-null  object 
 7   Delivery                                            21321 non-null  object 
 8   Distance                                            21321 non-null  object 


Restaurant ID      Order ID  Bill subtotal  Packaging charges  \
count   2.132100e+04  2.132100e+04   21321.000000       21321.000000   
mean    2.074413e+07  6.354622e+09     750.076838          32.564592   
std     2.447193e+05  1.230263e+08     498.759428          22.235898   
min     2.032061e+07  6.086767e+09      50.000000           0.000000   
25%     2.063570e+07  6.250751e+09     459.000000          18.450000   
50%     2.065987e+07  6.357715e+09     629.000000          28.450000   
75%     2.088265e+07  6.456827e+09     899.000000          39.950000   
max     2.152306e+07  6.573392e+09   16080.000000         603.000000   

       Restaurant discount (Promo)  \
count                 21321.000000   
mean                     65.091816   
std                      85.401604   
min                       0.000000   
25%                       0.000000   
50%                      80.000000   
75%                     100.000000   
max                    4020.000000   

       Restaurant discount (Flat offs, Freebies & others)  Gold discount  \
count                                       21321.000000    21321.000000   
mean                                           31.795058        0.099128   
std                                           131.487091        3.264261   
min                                             0.000000        0.000000   
25%                                             0.000000        0.000000   
50%                                             0.000000        0.000000   
75%                                             0.000000        0.000000   
max                                          7787.000000      280.100000   

       Brand pack discount         Total       Rating  \
count         21321.000000  21321.000000  2491.000000   
mean              3.039324    682.616113     4.356885   
std              17.070780    465.313977     1.181472   
min               0.000000     52.500000     1.000000   
25%               0.000000    387.450000     4.000000   
50%               0.000000    597.450000     5.000000   
75%               0.000000    837.900000     5.000000   
max             554.800000  12663.000000     5.000000   

       Restaurant compensation (Cancellation)  Restaurant penalty (Rejection)  \
count                              133.000000                             3.0   
mean                               356.409549                             0.0   
std                                328.128240                             0.0   
min                                 83.580000                             0.0   
25%                                191.950000                             0.0   
50%                                272.580000                             0.0   
75%                                397.840000                             0.0   
max                               3236.980000                             0.0   

       KPT duration (minutes)  Rider wait time (minutes)  
count            21026.000000               21153.000000  
mean                17.332960                   4.825070  
std                  6.283388                   4.982591  
min                  0.000000                   0.100000  
25%                 13.380000                   1.000000  
50%                 16.330000                   3.100000  
75%                 20.050000                   7.400000  
max                 90.870000                  73.800000

We are provided with 21,321 food delivery order records, across 11,607 unique customers. Each record includes order details such as date and time, customer ID, city, bill amount, and discounts applied. The data also contains additional fields on ratings, cancellations, and timings, which are not relevant for our first-order GMV analysis.

##  Missing Value Handling & Cleaning

To ensure data quality, we verified that three critical columns are complete: `Customer ID`, `Order Placed At`, and `Total`. These are essential for identifying first orders and calculating GMV. While several non-critical columns contain missing values—such as ratings, reviews, cancellation reasons, and operational timings—these are not required for the current A/B test scope.

In [8]:
# Check missing value counts
missing = df.isnull().sum().sort_values(ascending=False)
print(missing[missing > 0])

Restaurant penalty (Rejection)            21318
Restaurant compensation (Cancellation)    21188
Cancellation / Rejection reason           21135
Review                                    21025
Customer complaint tag                    20852
Instructions                              20601
Rating                                    18830
Discount construct                         5498
KPT duration (minutes)                      295
Rider wait time (minutes)                   168
dtype: int64


No rows needed to be dropped for missing values in required fields. The `Order Placed At` column was converted from string to datetime format for accurate time-based sorting.

In [9]:
# Drop rows missing critical fields
df = df.dropna(subset=['Customer ID', 'Order Placed At', 'Total'])

# Convert Order Placed At to datetime
df['Order Placed At'] = pd.to_datetime(df['Order Placed At'], format='%I:%M %p, %B %d %Y')

# Check after cleaning
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21321 entries, 0 to 21320
Data columns (total 29 columns):
 #   Column                                              Non-Null Count  Dtype         
---  ------                                              --------------  -----         
 0   Restaurant ID                                       21321 non-null  int64         
 1   Restaurant name                                     21321 non-null  object        
 2   Subzone                                             21321 non-null  object        
 3   City                                                21321 non-null  object        
 4   Order ID                                            21321 non-null  int64         
 5   Order Placed At                                     21321 non-null  datetime64[ns]
 6   Order Status                                        21321 non-null  object        
 7   Delivery                                            21321 non-null  object        
 8   Distan

## First Order Identification

To isolate each customer’s first-ever order, the dataset was sorted by `Customer ID` and `Order Placed At`. We then assigned an incremental order number to each transaction per user. Only orders where `order_num == 1` were retained, resulting in a dataset of 11,607 unique first orders.

In [10]:
# Sort orders by customer and order time
df_sorted = df.sort_values(['Customer ID', 'Order Placed At'])

# Assign order number per customer
df_sorted['order_num'] = df_sorted.groupby('Customer ID').cumcount() + 1

# Extract first orders only
df_first = df_sorted[df_sorted['order_num'] == 1].copy()

# Check results
print(f"Unique customers: {df['Customer ID'].nunique()}")
print(f"First orders found: {df_first.shape[0]}")

# Sanity check: preview first orders
df_first[['Customer ID', 'Order Placed At', 'Total']].head()

Unique customers: 11607
First orders found: 11607


Customer ID     Order Placed At  \
930    000285ae83ecf06a92b936d4f5b74342edb0e1940e1f00... 2024-09-03 21:41:00   
1173   00062fa202370fdd3076b794ec9358f36e35469bb8e5ff... 2024-09-08 17:43:00   
18086  001ab5fc3ee158b4d22e106897cee9b355e6eed50f163e... 2025-01-10 01:38:00   
15335  001fdf2511dd137361424c9c15ea54774476691ceec5fa... 2024-12-13 22:26:00   
6370   002afbd83626f0c699892fcb07b2ddf3858911482647b6... 2024-10-11 18:50:00   

         Total  
930     522.90  
1173    628.95  
18086  1332.40  
15335  1352.40  
6370    796.95

## Simulate A/B/C/D Groups

A simulated A/B test was created by randomly assigning each first-order customer to one of four discount groups: A (0%), B (10%), C (20%), or D (30%). The random seed was fixed to ensure reproducibility. This allows us to evaluate the relative impact of different discount levels on GMV.

In [11]:
# Set random seed for reproducibility
np.random.seed(42)

# Create random group assignment
customers = df_first['Customer ID'].unique()
group_df = pd.DataFrame({
    'Customer ID': customers,
    'group': np.random.choice(['A','B','C','D'], size=len(customers))
})

# Merge group assignment into first orders dataframe
df_first = df_first.merge(group_df, on='Customer ID', how='left')

# Map to readable labels
discount_map = {'A': '0%', 'B': '10%', 'C': '20%', 'D': '30%'}
df_first['group_label'] = df_first['group'].map(discount_map)

# Sanity check: show group counts
df_first['group_label'].value_counts(normalize=True)

group_label
0%     0.253640
10%    0.250625
30%    0.248988
20%    0.246748
Name: proportion, dtype: float64

## EDA Summary
The dataset is now fully prepared for A/B test analysis. We have identified 11,607 first orders, assigned customers to balanced discount groups, and validated all critical fields. This notebook concludes the data preparation phase. 